# Numbers and more in Python!

In this lecture, we will learn about numbers in Python and how to use them.

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

## Types of numbers

Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

Throughout this course we will be mainly working with integers or simple float number types.

Here is a table of the two main types we will spend most of our time working with some examples:

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

 
 
Now let's start with some basic arithmetic.

### Basic Arithmetic

In [1]:
# Addition
2+1

3

In [4]:
# Subtraction

2-1

1

In [ ]:
# Multiplication
2*2

4

In [ ]:
# Division
7/4

1.75

In [3]:
# Floor Division
7//4

1

**Whoa! What just happened? Last time I checked, 7 divided by 4 equals 1.75 not 1!**

The reason we get this result is because we are using "*floor*" division. The // operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

**So what if we just want the remainder after division?**

In [5]:
# Modulo
7%4

3

4 goes into 7 once, with a remainder of 3. The % operator returns the remainder after division.

### Arithmetic continued

In [6]:
# Powers
2**3

8

In [7]:
# Can also do roots this way
4**0.5

2.0

In [8]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [ ]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

## EXERCISE 1

**Answer these 3 questions without typing code. Then type code to check your answer.**

    What is the value of the expression 4 * (6 + 5)
    
    What is the value of the expression 4 * 6 + 5 
    
    What is the value of the expression 4 + 6 * 5 

In [9]:
#TO DO
4 * (6 + 5)

44

In [10]:
4 * 6 + 5 

29

In [11]:
4 + 6 * 5 

34

# EXERCISE 2

**Write the equation $\big(60 + \frac{10^2}{4} * 7\big) - \sqrt{169}$ ---> The solution should be $222$**

In [12]:
#TO DO
(60 + (10 ** 2) / 4 * 7) - 169 **0.5

222.0

## Variable Assignments

Now that we've seen how to use numbers in Python as a calculator let's see how we can assign names and create variables.

We use a single equals sign to assign labels to variables. Let's see a few examples of how we can do this.

In [20]:
# Let's create an object called "a" and assign it the number 5
a = 15
print(a)

15


Now if I call *a* in my Python script, Python will treat it as the number 5.

In [24]:
# Adding the objects
a+a

20

What happens on reassignment? Will Python let us write it over?

In [22]:
# Reassignment
a = 10

In [25]:
# Check
a

10

Yes! Python allows you to write over assigned variable names. We can also use the variables themselves when doing the reassignment. Here is an example of what I mean:

In [28]:
b = 23

In [29]:
a + b

43

In [30]:
# Use A to redefine A
a = a + a

In [31]:
# Check 
a

40

The names you use when creating these labels need to follow a few rules:

* names can not start with a number
* names can not contain spaces, use _ intead
* names can not contain any of these symbols:

      :'",<>/?|\!@#%^&*~-+
       
* it's considered best practice ([PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) that names are lowercase with underscores
* avoid using Python built-in keywords like `list` and `str`
* avoid using the single characters `l` (lowercase letter el), `O` (uppercase letter oh) and `I` (uppercase letter eye) as they can be confused with `1` and `0`

Using variable names can be a very useful way to keep track of different variables in Python. For example:

In [33]:
# Use object names to keep better track of what's going on in your code!
my_income = 100

tax_rate = 0.1

my_taxes = my_income * tax_rate

In [34]:
# Show my taxes!
my_taxes

10.0

So what have we learned? We learned some of the basics of numbers in Python. We also learned how to do arithmetic and use Python as a basic calculator. We then wrapped it up with learning about Variable Assignment in Python.

Up next we'll learn about Strings!

# EXERCISE 3

**assign 3 to the variable a. assign 5 to the variable b. How can you invert the assignments so that a = 5 and b = 3?**

In [39]:
a = 3
b = 5

old_a = a

a = b
b = old_a

print(a,b)

5 3


In [ ]:
# TO DO
a = 3
b = 5

c = a
a = b
b = c

# EXERCISE 4
**What is the *type* of the result of the expression 3 + 1.5 + 4?**

In [40]:
#TO DO
type(3 + 1.5 + 4)

float

In [41]:
v = 3 + 1.5 + 4
type(v)

float

## Dynamic Typing (OPT)

Python uses *dynamic typing*, meaning you can reassign variables to different data types. This makes Python very flexible in assigning data types; it differs from other languages that are *statically typed*.

In [62]:
my_dogs = 2.3
my_dogs
type(my_dogs)

float

In [45]:
my_dogs = ['Sammy', 'Frankie', 'Sammy']
my_dogs
type(my_dogs)

list

### Pros and Cons of Dynamic Typing
#### Pros of Dynamic Typing
* very easy to work with
* faster development time

#### Cons of Dynamic Typing
* may result in unexpected bugs!
* you need to be aware of `type()`

## Determining variable type with `type()`
You can check what type of object is assigned to a variable using Python's built-in `type()` function. Common data types include:
* **int** (for integer)
* **float**
* **str** (for string)
* **list**
* **tuple**
* **dict** (for dictionary)
* **set**
* **bool** (for Boolean True/False)

In [46]:
type(a)

int

In [60]:
a = 1.0
type(a)

float

In [61]:
a = bool(a)
print(a)

True


In [54]:
type(a)


bool